File: api/src/main/java/org/pmiops/workbench/firecloud/FirecloudRetryHandler.java
package org.pmiops.workbench.firecloud;

import jakarta.inject.Provider;
import jakarta.servlet.http.HttpServletResponse;
import java.net.SocketTimeoutException;
import java.util.logging.Logger;
import org.broadinstitute.dsde.workbench.client.sam.api.TermsOfServiceApi;
import org.pmiops.workbench.exceptions.ExceptionUtils;
import org.pmiops.workbench.exceptions.WorkbenchException;
import org.pmiops.workbench.terra.TerraServiceRetryHandler;
import org.pmiops.workbench.utils.ResponseCodeRetryPolicy;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.retry.backoff.BackOffPolicy;
import org.springframework.stereotype.Service;

@Service
public class FirecloudRetryHandler extends TerraServiceRetryHandler<ApiException> {

  private static final Logger logger = Logger.getLogger(FirecloudRetryHandler.class.getName());

  private static class FirecloudRetryPolicy extends ResponseCodeRetryPolicy {

    public FirecloudRetryPolicy() {
      super("Firecloud API");
    }

    @Override
    protected int getResponseCode(Throwable lastException) {
      if (lastException instanceof ApiException) {
        return ((ApiException) lastException).getCode();
      }
      if (lastException instanceof SocketTimeoutException) {
        return HttpServletResponse.SC_GATEWAY_TIMEOUT;
      }
      return HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
    }

    @Override
    protected void logNoRetry(Throwable t, int responseCode) {
      if (t instanceof ApiException) {
        logger.log(
            getLogLevel(responseCode),
            String.format(
                "Exception calling Firecloud API with response: %s",
                ((ApiException) t).getResponseBody()),
            t);
      } else {
        super.logNoRetry(t, responseCode);
      }
    }
  }

  @Autowired
  public FirecloudRetryHandler(
      BackOffPolicy backoffPolicy, Provider<TermsOfServiceApi> termsOfServiceApiProvider) {
    super(
        backoffPolicy,
        new FirecloudRetryPolicy(),
        termsOfServiceApiProvider,
        ExceptionUtils::convertFirecloudException);
  }

  @Override
  protected WorkbenchException convertException(ApiException exception) {
    return convertTerraException(exception, exception.getCode());
  }
}

File: api/src/main/java/org/pmiops/workbench/firecloud/FirecloudRetryHandler.kt
package org.pmiops.workbench.firecloud

import java.net.SocketTimeoutException
import java.util.logging.Logger
import javax.servlet.http.HttpServletResponse
import org.pmiops.workbench.config.RetryConfig
import org.pmiops.workbench.exceptions.ExceptionUtils
import org.pmiops.workbench.exceptions.WorkbenchException
import org.pmiops.workbench.utils.RetryHandler
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.retry.backoff.BackOffPolicy
import org.springframework.stereotype.Service

@Service
class FirecloudRetryHandler @Autowired
constructor(backoffPolicy: BackOffPolicy) : RetryHandler<ApiException>(backoffPolicy, FirecloudRetryPolicy()) {

    private class FirecloudRetryPolicy : RetryConfig.ResponseCodeRetryPolicy("Firecloud API") {

        override fun getResponseCode(lastException: Throwable): Int {
            if (lastException is ApiException) {
                return (lastException as ApiException).getCode()
            }
            return if (lastException is SocketTimeoutException) {
                HttpServletResponse.SC_GATEWAY_TIMEOUT
            } else HttpServletResponse.SC_INTERNAL_SERVER_ERROR
        }

        override fun logNoRetry(t: Throwable, responseCode: Int) {
            if (t is ApiException) {
                logger.log(
                        getLogLevel(responseCode),
                        String.format(
                                "Exception calling Firecloud API with response: %s",
                                (t as ApiException).getResponseBody()),
                        t)
            } else {
                super.logNoRetry(t, responseCode)
            }
        }
    }

    override fun convertException(exception: ApiException): WorkbenchException {
        return ExceptionUtils.convertFirecloudException(exception)
    }

    companion object {

        private val logger = Logger.getLogger(FirecloudRetryHandler::class.java.name)
    }
}
